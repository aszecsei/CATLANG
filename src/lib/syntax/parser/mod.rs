use std::rc::Rc;
use crate::syntax::ast;
use crate::syntax::lexer;
use crate::syntax::token;

use log::{debug, error};

use crate::syntax::error::Error;
use crate::syntax::context::Context;

pub struct Parser<'a> {
    fname: &'a str,
    errors: Vec<Error>,
    scanner: lexer::StringReader<'a>,
    current_scope: Rc<ast::Scope>,
    top_scope: Rc<ast::Scope>,
}

impl<'a> Parser<'a> {
    pub fn parse_file(fname: &'a str, src: &'a str, context: &'a mut Context) -> ast::Block {
        let mut p = Parser::new(fname, src, None, context);
        let res = p.parse_block();
        // TODO: Kinder error logging (i.e. hide > 10 errors)
        for e in p.errors {
            error!("{}", e.get_msg());
        }
        res
    }

    fn expect(&mut self, tok: token::Token) -> Result<(), Error> {
        let r = {
            let lexeme = self.scanner.token().tok;

            if lexeme != tok {
                Err(Error::new(format!(
                    "Expected '{}' but got '{}'",
                    tok, lexeme
                )))
            } else {
                Ok(())
            }
        };
        self.next();
        r
    }

    fn new(fname: &'a str, src: &'a str, scope: Option<ast::Scope>, context: &'a mut Context) -> Self {
        let scope = match scope {
            Some(s) => s,
            None => ast::Scope::new(None),
        };
        let rc_scope = Rc::new(scope);

        let source_file = {
            let mut source_map = context.get_source_map();
            let source_file = source_map.add_file(String::from(fname), String::from(src));
            source_file
        };

        let mut p = Parser {
            fname,
            errors: vec![],
            scanner: lexer::StringReader::new(context, source_file.clone(), src),
            current_scope: rc_scope.clone(),
            top_scope: rc_scope,
        };
        p.next();
        p
    }

    fn next(&mut self) {
        self.scanner.next();
    }

    fn open_scope(&mut self) {
        self.current_scope = Rc::new(ast::Scope::new(Some(self.current_scope.clone())));
    }

    fn close_scope(&mut self) {
        self.current_scope = self.current_scope.parent.clone().unwrap(); // TODO: Unwrap
    }

    fn is_declaration_starter(t: token::Token) -> bool {
        t == token::Token::Export
            || t == token::Token::Const
            || t == token::Token::Type
            || t == token::Token::Let
            || t == token::Token::Function
            || t == token::Token::Struct
            || t == token::Token::Enum
    }

    fn parse_block(&mut self) -> ast::Block {
        let mut elements = vec![];
        let mut sp = self.scanner.token().sp;
        loop {
            match self.scanner.token().tok {
                token::Token::EOF => break,
                token::Token::RCurlyB => break,
                token::Token::Semicolon => {
                    self.next();
                    continue;
                }
                tok => {
                    debug!("Parsing start of block element {:?}", tok);
                    sp = sp.merge(self.scanner.token().sp);

                    if Parser::is_declaration_starter(tok) {
                        let declaration = self.parse_declaration();
                        match declaration {
                            Err(e) => self.errors.push(e),
                            Ok(d) => elements.push(ast::BlockElement::Declaration(d)),
                        }
                    } else {
                        let statement = self.parse_statement();
                        match statement {
                            Err(e) => self.errors.push(e),
                            Ok(s) => elements.push(ast::BlockElement::Statement(s)),
                        }
                    }
                }
            }
        }
        ast::Block {
            elements,
            span: sp,
        }
    }

    fn parse_declaration(&mut self) -> Result<ast::Declaration, Error> {
        match self.scanner.token().tok {
            token::Token::Export => {
                self.next();
                let d = self.parse_declarator()?;
                Ok(ast::Declaration {
                    is_exported: true,
                    declarator: d,
                })
            }
            _ => {
                let d = self.parse_declarator()?;
                Ok(ast::Declaration {
                    is_exported: false,
                    declarator: d
                })
            }
        }
    }

    fn parse_declarator(&mut self) -> Result<ast::Declarator, Error> {
        match self.scanner.token().tok {
            token::Token::Const => Ok(ast::Declarator::ConstantDeclarator(
                self.parse_const_declarator()?,
            )),
            token::Token::Type => Ok(ast::Declarator::TypeDeclarator(
                self.parse_type_declarator()?,
            )),
            token::Token::Let => Ok(ast::Declarator::VariableDeclarator(
                self.parse_variable_declarator()?,
            )),
            token::Token::Function => Ok(ast::Declarator::FunctionDeclarator(
                self.parse_function_declarator()?,
            )),
            token::Token::Struct => Ok(ast::Declarator::StructDeclarator(
                self.parse_struct_declarator()?,
            )),
            token::Token::Enum => Ok(ast::Declarator::EnumDeclarator(
                self.parse_enum_declarator()?,
            )),
            tok => Err(Error::new(String::from(format!(
                "Expected a declarator, got {}",
                tok
            )))),
        }
    }

    fn parse_const_declarator(&mut self) -> Result<ast::ConstantDeclarator, Error> {
        self.expect(token::Token::Const)?;
        let identifier = self.parse_identifier()?;
        self.expect(token::Token::Assign)?;
        let expression = self.parse_expression()?;

        Ok(ast::ConstantDeclarator {
            identifier,
            expression
        })
    }

    fn parse_type_declarator(&mut self) -> Result<ast::TypeDeclarator, Error> {
        self.expect(token::Token::Type)?;
        let identifier = self.parse_identifier()?;
        self.expect(token::Token::Equals)?;
        let assigned_type = self.parse_type()?;

        Ok(ast::TypeDeclarator {
            identifier,
            type_expression: assigned_type,
        })
    }

    fn parse_variable_declarator(&mut self) -> Result<ast::VariableDeclarator, Error> {
        self.expect(token::Token::Let)?;
        let identifier = self.parse_identifier()?;

        let type_expression = match self.scanner.token().tok {
            token::Token::Colon => {
                self.next();
                Some(self.parse_type()?)
            }
            _ => None,
        };

        let expression = match self.scanner.token().tok {
            token::Token::Assign => {
                self.next();
                Some(self.parse_expression()?)
            }
            _ => None,
        };

        Ok(ast::VariableDeclarator {
            identifier,
            type_expression,
            expression
        })
    }

    fn parse_function_declarator(&mut self) -> Result<ast::FunctionDeclarator, Error> {
        self.expect(token::Token::Function)?;
        let identifier = self.parse_identifier()?;
        let parameters = self.parse_formal_parameter_list()?;
        self.expect(token::Token::Arrow)?;

        let return_type = match self.scanner.token().tok {
            token::Token::LCurlyB => None,
            token::Token::Arrow => {
                self.next();
                Some(self.parse_type()?)
            }
            tok => return Err(Error::new(String::from(format!("Expected either a return type or function start but got {}", tok))))
        };
        self.expect(token::Token::LCurlyB)?;
        let block = self.parse_block();
        self.expect(token::Token::RCurlyB)?;

        Ok(ast::FunctionDeclarator {
            identifier,
            parameters,
            return_type,
            block,
        })
    }

    fn parse_formal_parameter_list(&mut self) -> Result<Vec<ast::Parameter>, Error> {
        let mut params = vec![];
        self.expect(token::Token::LParen)?;

        match self.scanner.token().tok {
            token::Token::RParen => {
                self.next();
                return Ok(params);
            }
            _ => params.push(self.parse_parameter()?),
        }

        loop {
            match self.scanner.token().tok {
                token::Token::RParen => break,
                token::Token::Comma => {
                    self.next();
                    params.push(self.parse_parameter()?);
                }
                tok => return Err(Error::new(format!("Unexpected token {:?}", tok))),
            }
        }
        self.expect(token::Token::RParen)?;

        Ok(params)
    }

    fn parse_parameter(&mut self) -> Result<ast::Parameter, Error> {
        let is_const = match self.scanner.token().tok {
            token::Token::Const => {
                self.next();
                true
            },
            _ => false
        };
        let identifier = self.parse_identifier()?;
        self.expect(token::Token::Colon)?;
        let type_expression = self.parse_type()?;

        Ok(ast::Parameter {
            is_const,
            identifier,
            type_expression,
        })
    }

    fn parse_struct_declarator(&mut self) -> Result<ast::StructDeclarator, Error> {
        self.expect(token::Token::Struct)?;
        let identifier = self.parse_identifier()?;
        let members = self.parse_struct_members()?;

        Ok(ast::StructDeclarator {
            identifier,
            is_soa: false,
            members,
        })
    }

    fn parse_struct_members(&mut self) -> Result<Vec<ast::StructMember>, Error> {
        let mut members = vec![];
        self.expect(token::Token::LCurlyB)?;

        loop {
            match self.scanner.token().tok {
                token::Token::RCurlyB => break,
                _ => {
                    members.push(self.parse_struct_member()?);
                    self.expect(token::Token::Semicolon)?;
                }
            }
        }
        self.expect(token::Token::RCurlyB)?;

        Ok(members)
    }

    fn parse_struct_member(&mut self) -> Result<ast::StructMember, Error> {
        let identifier = self.parse_identifier()?;
        self.expect(token::Token::Colon)?;

        let is_owned = match self.scanner.token().tok {
            token::Token::Owned => {
                self.next();
                true
            },
            _ => false
        };

        let type_expression = self.parse_type()?;

        let has_default_value = self.scanner.token().tok == token::Token::Assign;
        let default_value = if has_default_value {
            self.next();
            Some(self.parse_expression()?)
        } else {
            None
        };

        Ok(ast::StructMember {
            identifier,
            is_owned,
            type_expression,
            default_value,
        })
    }

    fn parse_enum_declarator(&mut self) -> Result<ast::EnumDeclarator, Error> {
        self.expect(token::Token::Enum)?;
        let identifier = self.parse_identifier()?;
        let values = self.parse_enum_value_list()?;

        Ok(ast::EnumDeclarator { identifier, values })
    }

    fn parse_enum_value_list(&mut self) -> Result<Vec<ast::Ident>, Error> {
        let mut values = vec![];
        self.expect(token::Token::LCurlyB)?;

        loop {
            match self.scanner.token().tok {
                token::Token::RCurlyB => break,
                _ => {
                    values.push(self.parse_identifier()?);
                    self.expect(token::Token::Semicolon)?;
                }
            }
        }
        self.expect(token::Token::RCurlyB)?;

        Ok(values)
    }

    fn parse_statement(&mut self) -> Result<ast::Statement, Error> {
        match self.scanner.token().tok {
            token::Token::Import => Ok(ast::Statement::ImportStatement(self.parse_import()?)),
            token::Token::LCurlyB => Ok(ast::Statement::InnerBlock(self.parse_inner_block()?)),
            token::Token::If => Ok(ast::Statement::IfStatement(self.parse_if()?)),
            token::Token::For => Ok(ast::Statement::LoopStatement(self.parse_loop()?)),
            token::Token::While => Ok(ast::Statement::LoopStatement(self.parse_loop()?)),
            token::Token::Do => Ok(ast::Statement::LoopStatement(self.parse_loop()?)),
            token::Token::Break => Ok(ast::Statement::JumpStatement(self.parse_jump()?)),
            token::Token::Continue => Ok(ast::Statement::JumpStatement(self.parse_jump()?)),
            token::Token::Return => Ok(ast::Statement::JumpStatement(self.parse_jump()?)),
            _ => Ok(ast::Statement::Expression(self.parse_expression()?)),
        }
    }

    fn parse_import(&mut self) -> Result<ast::ImportStatement, Error> {
        self.next();
        Err(Error::new(String::from("Imports not implemented")))
    }

    fn parse_inner_block(&mut self) -> Result<ast::Block, Error> {
        self.next();
        Err(Error::new(String::from("Inner blocks not implemented")))
    }

    fn parse_if(&mut self) -> Result<ast::IfStatement, Error> {
        self.next();
        Err(Error::new(String::from("If statements not implemented")))
    }

    fn parse_loop(&mut self) -> Result<ast::LoopStatement, Error> {
        self.next();
        Err(Error::new(String::from("Loops not implemented")))
    }

    fn parse_jump(&mut self) -> Result<ast::JumpStatement, Error> {
        self.next();
        Err(Error::new(String::from("Jumps not implemented")))
    }

    fn parse_expression(&mut self) -> Result<ast::Expression, Error> {
        self.parse_assignment_expression()
    }

    // Right-associative
    fn parse_assignment_expression(&mut self) -> Result<ast::Expression, Error> {
        let assignment_operators = vec![
            token::Token::Assign,
            token::Token::AddAssign,
            token::Token::SubAssign,
            token::Token::MulAssign,
            token::Token::QuoAssign,
            token::Token::ModAssign,
            token::Token::BitOrAssign,
            token::Token::BitAndAssign,
            token::Token::XorAssign,
            token::Token::ShiftLAssign,
            token::Token::ShiftRAssign,
        ];
        let lhs = self.parse_conditional_expression()?;
        let tok = self.scanner.token().tok;
        if assignment_operators.contains(&tok) {
            let assignment_operator = tok;
            self.next();
            let rhs = self.parse_expression()?;
            Ok(ast::Expression::BinaryExpression(ast::BinaryExpression {
                left_hand_side: Box::new(lhs),
                operator: assignment_operator,
                right_hand_side: Box::new(rhs),
            }))
        } else {
            Ok(lhs)
        }
    }

    // Right-associative
    fn parse_conditional_expression(&mut self) -> Result<ast::Expression, Error> {
        let condition = self.parse_comparing_expression()?;
        let tok = self.scanner.token().tok;
        if tok == token::Token::Question {
            self.next();
            let true_value = Box::new(self.parse_expression()?);
            self.expect(token::Token::Colon)?;
            let false_value = Box::new(self.parse_expression()?);
            Ok(ast::Expression::TernaryExpression(ast::TernaryExpression {
                condition: Box::new(condition),
                true_value,
                false_value,
            }))
        } else {
            Ok(condition)
        }
    }

    // Right-associative
    fn parse_comparing_expression(&mut self) -> Result<ast::Expression, Error> {
        let comparing_operators = vec![
            token::Token::Equals,
            token::Token::GreaterThan,
            token::Token::GreaterThanEquals,
            token::Token::NotEquals,
            token::Token::LessThan,
            token::Token::LessThanEquals,
            token::Token::In,
        ];

        let lhs = self.parse_comparand()?;
        let tok = self.scanner.token().tok;
        if comparing_operators.contains(&tok) {
            let comparing_operator = tok;
            self.next();
            let rhs = self.parse_expression()?;
            Ok(ast::Expression::BinaryExpression(ast::BinaryExpression {
                left_hand_side: Box::new(lhs),
                operator: comparing_operator,
                right_hand_side: Box::new(rhs),
            }))
        } else {
            Ok(lhs)
        }
    }

    // Left-associative
    fn parse_comparand(&mut self) -> Result<ast::Expression, Error> {
        let adding_operators = vec![
            token::Token::Add,
            token::Token::Sub,
            token::Token::Or,
            token::Token::BitOr,
        ];

        let lhs = self.parse_term()?;
        let mut expr = lhs;

        loop {
            let tok = self.scanner.token().tok;
            if adding_operators.contains(&tok) {
                let adding_operator = tok;
                self.next();

                let rhs = self.parse_term()?;
                expr = ast::Expression::BinaryExpression(ast::BinaryExpression {
                    left_hand_side: Box::new(expr),
                    operator: adding_operator,
                    right_hand_side: Box::new(rhs),
                });
            } else {
                break;
            }
        }
        Ok(expr)
    }

    // Left-associative
    fn parse_term(&mut self) -> Result<ast::Expression, Error> {
        let multiplying_operators = vec![
            token::Token::Mul,
            token::Token::Quo,
            token::Token::And,
            token::Token::BitAnd,
        ];

        let lhs = self.parse_factor()?;
        let mut expr = lhs;

        loop {
            let tok = self.scanner.token().tok;
            if multiplying_operators.contains(&tok) {
                let multiplying_operator = tok;
                self.next();
                let rhs = self.parse_factor()?;

                expr = ast::Expression::BinaryExpression(ast::BinaryExpression {
                    left_hand_side: Box::new(expr),
                    operator: multiplying_operator,
                    right_hand_side: Box::new(rhs),
                });
            } else {
                break;
            }
        }
        Ok(expr)
    }

    // Left-associative
    fn parse_factor(&mut self) -> Result<ast::Expression, Error> {
        let shifting_operators = vec![token::Token::ShiftL, token::Token::ShiftR];

        let lhs = self.parse_unary()?;
        let mut expr = lhs;

        loop {
            let tok = self.scanner.token().tok;
            if shifting_operators.contains(&tok) {
                let shifting_operator = tok;
                self.next();

                let rhs = self.parse_unary()?;
                expr = ast::Expression::BinaryExpression(ast::BinaryExpression {
                    left_hand_side: Box::new(expr),
                    operator: shifting_operator,
                    right_hand_side: Box::new(rhs),
                });
            } else {
                break;
            }
        }
        Ok(expr)
    }

    fn parse_unary(&mut self) -> Result<ast::Expression, Error> {
        let prefix_unary_operators = vec![
            token::Token::Not,
            token::Token::Sub,
            token::Token::Increment,
            token::Token::Decrement,
        ];

        let postfix_unary_operators = vec![token::Token::Increment, token::Token::Decrement];

        let tok = self.scanner.token().tok;
        if prefix_unary_operators.contains(&tok) {
            self.next();
            let rhs = self.parse_unary()?;
            Ok(ast::Expression::UnaryPrefixExpression(
                ast::UnaryExpression {
                    operator: tok,
                    expression: Box::new(rhs),
                },
            ))
        } else {
            let mut expr = self.parse_primary_expression()?;
            loop {
                let tok = self.scanner.token().tok;
                if postfix_unary_operators.contains(&tok) {
                    self.next();
                    expr = ast::Expression::UnaryPostfixExpression(ast::UnaryExpression {
                        operator: tok,
                        expression: Box::new(expr),
                    });
                } else {
                    break;
                }
            }
            Ok(expr)
        }
    }

    fn parse_primary_expression(&mut self) -> Result<ast::Expression, Error> {
        let tok = self.scanner.token().tok;
        match tok {
            token::Token::Integer(_x) => {
                self.next();
                Ok(ast::Expression::PrimaryExpression(
                    ast::PrimaryExpression::Literal(tok),
                ))
            }
            token::Token::String(_x) => {
                self.next();
                Ok(ast::Expression::PrimaryExpression(
                    ast::PrimaryExpression::Literal(tok),
                ))
            }
            token::Token::Null => {
                self.next();
                Ok(ast::Expression::PrimaryExpression(
                    ast::PrimaryExpression::Null,
                ))
            }
            token::Token::LParen => {
                self.next(); // Consume left paren
                             // Determine if this is a lambda or a subexpression
                let x = self.scanner.peek().tok;
                if x == token::Token::Colon {
                    // TODO: Lambdas
                    Err(Error::new(String::from("Lambdas are not yet supported")))
                } else {
                    let r = self.parse_expression()?;
                    self.expect(token::Token::RParen)?;
                    Ok(ast::Expression::PrimaryExpression(
                        ast::PrimaryExpression::SubExpression(Box::new(r)),
                    ))
                }
            }
            _ => Ok(ast::Expression::PrimaryExpression(
                ast::PrimaryExpression::Reference(self.parse_reference()?),
            )),
        }
    }

    fn parse_reference(&mut self) -> Result<ast::Reference, Error> {
        let tok = self.scanner.token().tok;
        let r = match tok {
            token::Token::At => {
                self.next();
                ast::Reference::AddressOf(Box::new(self.parse_reference()?))
            }
            token::Token::Mul => {
                self.next();
                ast::Reference::Dereference(Box::new(self.parse_reference()?))
            }
            _ => ast::Reference::Ident(self.parse_identifier()?),
        };

        // TODO: member access
        // TODO: function call
        // TODO: constructor call
        // TODO: array reference
        // TODO: cast reference

        Ok(r)
    }

    fn parse_type(&mut self) -> Result<ast::TypeExpression, Error> {
        let mut type_expr = self.parse_unary_type()?;
        loop {
            let tok = self.scanner.token().tok;
            match tok {
                token::Token::BitOr => {
                    // Type union
                    self.next();
                    let other_type = self.parse_unary_type()?;
                    type_expr = ast::TypeExpression::TypeUnion(ast::TypeUnion {
                        first_type: Box::new(type_expr),
                        second_type: Box::new(other_type),
                    });
                }
                _ => {
                    break;
                }
            }
        }
        Ok(type_expr)
    }

    fn parse_unary_type(&mut self) -> Result<ast::TypeExpression, Error> {
        let tok = self.scanner.token().tok;
        let mut type_expr = match tok {
            token::Token::Mul => {
                self.next();
                ast::TypeExpression::PointerType(Box::new(self.parse_unary_type()?))
            }
            token::Token::LSquareB => {
                self.next();
                let tok = self.scanner.token().tok;
                match tok {
                    token::Token::DotDot => {
                        // Unsized array
                        self.next();
                        self.expect(token::Token::RSquareB)?;
                        let array_type = self.parse_unary_type()?;
                        ast::TypeExpression::UnsizedArrayType(Box::new(array_type))
                    }
                    _ => {
                        // Sized array
                        let array_size = self.parse_expression()?;
                        self.expect(token::Token::RSquareB)?;
                        let array_type = self.parse_unary_type()?;
                        ast::TypeExpression::SizedArrayType(ast::SizedArrayType {
                            size: array_size,
                            type_expression: Box::new(array_type),
                        })
                    }
                }
            }
            token::Token::Typeof => {
                self.next();
                let expr = self.parse_expression()?;
                ast::TypeExpression::TypeofExpression(expr)
            }
            token::Token::LParen => {
                self.next();
                let t = self.parse_type()?;
                self.expect(token::Token::RParen)?;
                t
            }
            _ => ast::TypeExpression::NamedType(self.parse_identifier()?),
        };
        loop {
            let tok = self.scanner.token().tok;
            match tok {
                token::Token::Question => {
                    // Optional type
                    self.next();
                    type_expr = ast::TypeExpression::OptionalType(Box::new(type_expr));
                }
                _ => {
                    break;
                }
            }
        }
        Ok(type_expr)
    }

    fn parse_identifier(&mut self) -> Result<ast::Ident, Error> {
        match self.scanner.token().tok {
            token::Token::Ident(sym) => {
                self.next();
                Ok(ast::Ident { name: sym })
            }
            tok => Err(Error::new(format!(
                "Expected identifier but got {}",
                tok
            ))),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_expect() {
        let mut context = Context::new();
        let mut p = Parser::new("test.cat", "; 21", None, &mut context);
        let res1 = p.expect(token::Token::Semicolon);
        let res2 = p.expect(token::Token::Semicolon);
        let res3 = p.expect(token::Token::Semicolon);

        assert!(res1.is_ok());
        assert!(res2.is_err());
        assert!(res3.is_err());
    }

    #[test]
    fn test_is_declaration_starter() {
        let r1 = Parser::is_declaration_starter(token::Token::Function);
        let r2 = Parser::is_declaration_starter(token::Token::Comma);

        assert!(r1);
        assert!(!r2);
    }
}