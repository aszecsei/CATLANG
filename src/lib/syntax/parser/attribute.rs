use toolshed::list::GrowableList;

use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn attribute_node(&mut self) -> Result<AttributeNode<'ast>> {
        unimplemented!()
    }
}
