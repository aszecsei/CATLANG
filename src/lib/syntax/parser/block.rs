use toolshed::list::GrowableList;

use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::module::is_declaration_starter;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn block_node(&mut self) -> Result<BlockNode<'ast>> {
        let start = self.loc().0;
        self.expect(Token::LCurlyB);
        let elements = GrowableList::new();
        loop {
            match self.current_token {
                Token::RCurlyB => break,
                Token::Semicolon => self.bump(),
                _ => elements.push(self.arena, self.block_element()?),
            }
        }
        let end = self.loc().1;
        self.expect(Token::RCurlyB);
        Ok(self.node_at(
            start,
            end,
            Block {
                elements: elements.as_list(),
            },
        ))
    }

    fn block_element(&mut self) -> Result<BlockElementNode<'ast>> {
        if self.current_token == Token::Import {
            let import = self.import()?;
            Ok(self.node_at(import.start, import.end, import))
        } else if is_declaration_starter(self.current_token) {
            let declaration = self.declaration_node()?;
            Ok(self.node_at(declaration.start, declaration.end, declaration))
        } else {
            let statement = self.statement_node()?;
            Ok(self.node_at(statement.start, statement.end, statement))
        }
    }
}
