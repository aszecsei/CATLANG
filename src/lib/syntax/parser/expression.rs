use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn expression_node(&mut self) -> Result<ExpressionNode<'ast>> {
        self.assignment_expression()
    }

    // Right-associative
    fn assignment_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.conditional_expression()?;
        let assignment = match self.current_token {
            Token::Assign => Some(AssignmentOperator::Plain),
            Token::AddAssign => Some(AssignmentOperator::Addition),
            Token::SubAssign => Some(AssignmentOperator::Subtraction),
            Token::MulAssign => Some(AssignmentOperator::Multiplication),
            Token::QuoAssign => Some(AssignmentOperator::Division),
            Token::ModAssign => Some(AssignmentOperator::Remainder),
            Token::BitOrAssign => Some(AssignmentOperator::BitOr),
            Token::OrAssign => Some(AssignmentOperator::LogicalOr),
            Token::BitAndAssign => Some(AssignmentOperator::BitAnd),
            Token::AndAssign => Some(AssignmentOperator::LogicalAnd),
            Token::XorAssign => Some(AssignmentOperator::BitXor),
            Token::ShiftLAssign => Some(AssignmentOperator::BitShiftLeft),
            Token::ShiftRAssign => Some(AssignmentOperator::BitShiftRight),
            Token::NullCoalesceAssign => Some(AssignmentOperator::NullCoalesce),
            _ => None,
        };
        if let Some(assignment_operator) = assignment {
            self.bump();
            let rhs = self.expression_node()?;
            Ok(self.node_at(
                lhs.start,
                rhs.end,
                AssignmentExpression {
                    left: lhs,
                    operator: assignment_operator,
                    right: rhs,
                },
            ))
        } else {
            Ok(lhs)
        }
    }

    // Right-associative
    fn conditional_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        let condition = self.comparing_expression()?;
        if self.eat(Token::Question) {
            let true_value = self.expression_node()?;
            self.expect(Token::Colon);
            let false_value = self.expression_node()?;
            Ok(self.node_at(
                condition.start,
                false_value.end,
                TernaryExpression {
                    condition,
                    when_true: true_value,
                    when_false: false_value,
                },
            ))
        } else {
            Ok(condition)
        }
    }

    // Right-associative
    fn comparing_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.comparand()?;
        let operator = match self.current_token {
            Token::Equals => Some(BinaryOperator::Equals),
            Token::GreaterThan => Some(BinaryOperator::GreaterThan),
            Token::GreaterThanEquals => Some(BinaryOperator::GreaterThanEquals),
            Token::NotEquals => Some(BinaryOperator::NotEquals),
            Token::LessThan => Some(BinaryOperator::LessThan),
            Token::LessThanEquals => Some(BinaryOperator::LessThanEquals),
            Token::In => Some(BinaryOperator::In),
            _ => None,
        };
        if let Some(binary_operator) = operator {
            self.bump();
            let rhs = self.expression_node()?;
            Ok(self.node_at(
                lhs.start,
                rhs.end,
                BinaryExpression {
                    left: lhs,
                    operator: binary_operator,
                    right: rhs,
                },
            ))
        } else {
            Ok(lhs)
        }
    }

    // Left-associative
    fn comparand(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.term()?;
        let mut expr = lhs;
        loop {
            let operator = match self.current_token {
                Token::Add => Some(BinaryOperator::Addition),
                Token::Sub => Some(BinaryOperator::Subtraction),
                Token::Or => Some(BinaryOperator::LogicalOr),
                Token::BitOr => Some(BinaryOperator::BitOr),
                _ => None,
            };
            if let Some(binary_operator) = operator {
                self.bump();
                let rhs = self.term()?;
                expr = self.node_at(
                    expr.start,
                    expr.end,
                    BinaryExpression {
                        left: lhs,
                        operator: binary_operator,
                        right: rhs,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    // Left-associative
    fn term(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.factor()?;
        let mut expr = lhs;
        loop {
            let operator = match self.current_token {
                Token::Mul => Some(BinaryOperator::Multiplication),
                Token::Quo => Some(BinaryOperator::Division),
                Token::And => Some(BinaryOperator::LogicalAnd),
                Token::BitAnd => Some(BinaryOperator::BitAnd),
                _ => None,
            };
            if let Some(binary_operator) = operator {
                self.bump();
                let rhs = self.factor()?;
                expr = self.node_at(
                    expr.start,
                    expr.end,
                    BinaryExpression {
                        left: lhs,
                        operator: binary_operator,
                        right: rhs,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    // Left-associative
    fn factor(&mut self) -> Result<ExpressionNode<'ast>> {
        let lhs = self.unary()?;
        let mut expr = lhs;
        loop {
            let operator = match self.current_token {
                Token::ShiftL => Some(BinaryOperator::BitShiftLeft),
                Token::ShiftR => Some(BinaryOperator::BitShiftRight),
                _ => None,
            };
            if let Some(binary_operator) = operator {
                self.bump();
                let rhs = self.unary()?;
                expr = self.node_at(
                    expr.start,
                    expr.end,
                    BinaryExpression {
                        left: lhs,
                        operator: binary_operator,
                        right: rhs,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    fn unary(&mut self) -> Result<ExpressionNode<'ast>> {
        let operator = match self.current_token {
            Token::Not => Some(PrefixOperator::LogicalNot),
            Token::BitNot => Some(PrefixOperator::BitNot),
            Token::Sub => Some(PrefixOperator::Minus),
            Token::Add => Some(PrefixOperator::Plus),
            Token::Increment => Some(PrefixOperator::Increment),
            Token::Decrement => Some(PrefixOperator::Decrement),
            _ => None,
        };
        if let Some(prefix_operator) = operator {
            let start = self.loc().0;
            self.bump();
            let rhs = self.unary()?;
            return Ok(self.node_at(
                start,
                rhs.end,
                PrefixExpression {
                    operator: prefix_operator,
                    operand: rhs,
                },
            ));
        }

        let mut expr = self.primary_expression()?;
        loop {
            let operator = match self.current_token {
                Token::Increment => Some(PostfixOperator::Increment),
                Token::Decrement => Some(PostfixOperator::Decrement),
                Token::Not => Some(PostfixOperator::NullForgiving),
                _ => None,
            };
            if let Some(postfix_operator) = operator {
                let end = self.loc().1;
                self.bump();
                expr = self.node_at(
                    expr.start,
                    end,
                    PostfixExpression {
                        operator: postfix_operator,
                        operand: expr,
                    },
                );
            } else {
                break Ok(expr);
            }
        }
    }

    fn primary_expression(&mut self) -> Result<ExpressionNode<'ast>> {
        match self.current_token {
            Token::Integer(_) => Ok(self.node_from_slice(|s| Primitive::DecimalNumber(s))),
            Token::LiteralString => Ok(self.node_from_slice(|s| Primitive::String(s))),
            Token::Bool(b) => Ok(self.node_at_token(Primitive::Bool(b))),
            Token::Null => Ok(self.node_at_token(Primitive::Null)),
            Token::LParen => {
                self.expect(Token::LParen);
                if self.peek_token == Token::Colon {
                    // TODO: Lambdas
                    Err(Error::NotImplementedError)
                } else {
                    let inside = self.expression_node()?;
                    self.expect(Token::RParen);
                    Ok(inside)
                }
            }
            _ => self.reference(),
        }
    }

    fn reference(&mut self) -> Result<ExpressionNode<'ast>> {
        // TODO: Remove this exception
        #[allow(clippy::match_single_binding)]
        match self.current_token {
            // Token::At => Address-of reference
            // Token::Mul => Dereference
            // TODO: Member access
            // TODO: Function call
            // TODO: Constructor call
            // TODO: Array reference
            // TODO: Cast reference
            _ => {
                let ident = self.identifier_node()?;
                Ok(self.node_at(ident.start, ident.end, ident))
            }
        }
    }
}
