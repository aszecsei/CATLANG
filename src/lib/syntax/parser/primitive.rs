use crate::syntax::ast::*;
use crate::syntax::error::*;
use crate::syntax::lexer::Token;
use crate::syntax::parser::Parser;

impl<'ast> Parser<'ast> {
    pub fn identifier_node(&mut self) -> Result<IdentifierNode<'ast>> {
        let (start, end) = (self.current_span.start as u32, self.current_span.end as u32);
        let val = self.current_slice;
        self.expect(Token::Ident);
        Ok(self.node_at(start, end, val))
    }

    pub fn string_literal_node(&mut self) -> Result<StringLiteralNode<'ast>> {
        let (start, end) = (self.current_span.start as u32, self.current_span.end as u32);
        let val = self.current_slice;
        self.expect(Token::LiteralString);
        Ok(self.node_at(start, end, val))
    }
}
