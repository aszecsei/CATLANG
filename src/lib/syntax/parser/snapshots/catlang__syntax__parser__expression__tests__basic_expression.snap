---
source: src/lib/syntax/parser/expression.rs
expression: res
---
(1:11) BinaryExpression(
    BinaryExpression {
        left: (1:6) BinaryExpression(
            BinaryExpression {
                left: (1:2) PrimitiveExpression(
                    DecimalNumber(
                        "1",
                    ),
                ),
                operator: Addition,
                right: (5:6) PrimitiveExpression(
                    DecimalNumber(
                        "2",
                    ),
                ),
            },
        ),
        operator: Multiplication,
        right: (10:11) PrimitiveExpression(
            DecimalNumber(
                "3",
            ),
        ),
    },
)
